paasta_tools/marathon_dashboard.py

```diff
@@ -77,33 +79,48 @@ def create_marathon_dashboard(
-    for service_instance in instances:
-        service: str = service_instance[0]
-        instance: str = service_instance[1]
-        service_config: MarathonServiceConfig = load_marathon_service_config(
+
+    dashboard_links: Dict = system_paasta_config.get_dashboard_links()
+    marathon_links = dashboard_links.get(cluster, {}).get('Marathon RO')
+
+    # e.g. 'http://10.64.97.75:5052': 'http://marathon-norcal-prod.yelpcorp.com'
+    shard_url_to_marathon_link_dict: Dict[str, str] = {}
+    if isinstance(marathon_links, list):
+        # Sanity check and log error if necessary
+        if len(marathon_links) != len(marathon_servers.current):
+            log.error('len(marathon_links) != len(marathon_servers.current). This may be a cause of concern')
+        for shard_number, shard in enumerate(marathon_servers.current):
+            shard_url_to_marathon_link_dict[shard.url[0]] = marathon_links[shard_number]
+    elif isinstance(marathon_links, str):
+        # In this case, the shard url will be the same for every service instance
+        static_shard_url = marathon_links.split(' ')[0]
+        return {cluster: [{'service': si[0], 'instance': si[1], 'shard_url': static_shard_url} for si in instances]}
+
+    # Setup with service as key since will instantiate 1 PSCL per service
+    service_instances_dict: Dict[str, Set[str]] = defaultdict(set)
+    for si in instances:
+        service, instance = si[0], si[1]
+        service_instances_dict[service].add(instance)
+
+    for service, instance_set in service_instances_dict.items():
+        pscl = PaastaServiceConfigLoader(
             service=service,
-            instance=instance,
-            cluster=cluster,
-            load_deployments=False,
             soa_dir=soa_dir,
+            load_deployments=False,
         )
-        client: MarathonClient = marathon_clients.get_current_client_for_service(job_config=service_config)
-        dashboard_links: Dict = system_paasta_config.get_dashboard_links()
-        shard_url: str = client.servers[0]
-        if 'Marathon RO' in dashboard_links[cluster]:
-            marathon_links = dashboard_links[cluster]['Marathon RO']
-            if isinstance(marathon_links, list):
-                for shard_number, shard in enumerate(marathon_servers.current):
-                    if shard.url[0] == shard_url:
-                        shard_url = marathon_links[shard_number]
-            elif isinstance(marathon_links, str):
-                shard_url = marathon_links.split(' ')[0]
-        service_info: Marathon_Dashboard_Item = {
-            'service': service,
-            'instance': instance,
-            'shard_url': shard_url,
-        }
-        dashboard[cluster].append(service_info)
+        for marathon_service_config in pscl.instance_configs(cluster, MarathonServiceConfig):
+            if marathon_service_config.get_instance() in instance_set:
```

Comment:
I guess this is necessary because get_services_for_cluster is where the logic for ignoring instance names that start with `_` lives?

Reply:
Yea, I mean, meh. Why not just iterate over the pscl.instance_configs directly? Do we even need the `service_instances_dict` thingy?